File: 540-single-element-in-a-sorted-array/540-single-element-in-a-sorted-array.java
class Solution {
    public int singleNonDuplicate(int[] nums) {
      int n=nums.length;
        
        if(n==1){
            return nums[0];
        }else if(nums[0]!=nums[1]){
            return nums[0];
        }else if(nums[n-1]!=nums[n-2])return nums[n-1];
        
        
        int low=0;
        int high=nums.length-1;
        
        while(low<=high){
            int mid=(low+high)>>1;
            
            if(nums[mid]!=nums[mid-1] && nums[mid]!=nums[mid+1])return nums[mid];
            
            else if(nums[mid]==nums[mid-1]){
                int count=mid-low+1;
                
                if((count&1)==0){
                    low=mid+1;
                }else{
                    high=mid-2;
                }
                
            }else if(nums[mid]==nums[mid+1]){
                  int count=high-mid+1;
                
                if((count&1)==0){
                       high=mid-1;
                  
                }else{
                   low=mid+2;
                }
            }
        }
        
        return -1;
    }
}

File: 540-single-element-in-a-sorted-array/540-single-element-in-a-sorted-array.cpp
class Solution {
public:
    int singleNonDuplicate(vector<int>& nums) {
//         int low=0;
//         int high=nums.size()-2;
        
//         while(low<=high){
//             int mid=(low+high)>>1;
            
//             if(nums[mid]==nums[mid^1]){
//                 low=mid+1;
//             }else{
//                 high=mid-1;
//             }
//         }
        
//         return nums[low];
        
        
        int x=nums[0];
        for(int i=1;i<nums.size();i++){
            x^=nums[i];
        }
        
        return x;
        
    }
};

//for the left half
// first index -> even
//second index -> odd